.github/workflows/smoke-test.yaml

```diff
@@ -152,6 +160,93 @@ jobs:
+      matrix:
+        include:
+        - { id: opensuse-tumbleweed, image: "registry.opensuse.org/opensuse/tumbleweed:latest"}
+        - { id: opensuse-leap, image: "registry.opensuse.org/opensuse/leap:latest" }
+        - { id: debian, image: "debian:latest" }
+        - { id: fedora, image: "fedora:latest" }
+    runs-on: ubuntu-latest
+    container:
+      image: ${{ matrix.image }}
+      options: --privileged
+    steps:
+    - name: Install basic tools
+      if: contains(matrix.id, 'opensuse')
+      run: >-
+        zypper --non-interactive install
+        git-core tar
+    - uses: actions/checkout@08c6903cd8c0fde910a37f88322edcfb5dd907a8 # v5.0.0
+      with:
+        persist-credentials: false
+    - name: Set up user
+      run: |
+        useradd --create-home --user-group ci-user
+        export LOGS_DIR=$PWD/logs
+        export RD_LOGS_DIR=$LOGS_DIR/rd
+        echo "LOGS_DIR=$LOGS_DIR" >> "$GITHUB_ENV"
+        echo "RD_LOGS_DIR=$RD_LOGS_DIR" >> "$GITHUB_ENV"
+        mkdir -p $RD_LOGS_DIR
+        chown --recursive ci-user "$LOGS_DIR"
+    - uses: ./.github/actions/setup-environment
+    - name: Install Rancher Desktop from package
+      if: runner.os == 'Linux'
+      run: .github/workflows/smoke-test/install-from-repo.sh
+      env:
+        RD_VERSION: ${{ needs.download-artifacts.outputs.release-tag }}
+    - name: "openSUSE Workaround for #9145"
+      if: contains(matrix.id, 'opensuse')
+      run: >-
+        zypper --non-interactive install
+        qemu-img qemu-hw-display-virtio-gpu
+    - name: Run smoke test
+      shell: bash
+      run: |
+        inner_command=(
+          xvfb-run
+            --auto-servernum
+            --server-args='-screen 0 1280x960x24'
+          $PWD/.github/workflows/smoke-test/smoke-test.sh
+        )
+        sudo --user=ci-user --login --set-home --non-interactive \
+          /usr/bin/env --chdir=$PWD \
+            RD_DEBUG_ENABLED=1 RD_TEST=smoke RD_SKIP_INSTALL=true RD_LOGS_DIR=$RD_LOGS_DIR \
+          script \
+            --log-out $LOGS_DIR/repo-${{ matrix.id }}.log \
+            --return --command "${inner_command[*]@Q}"
+
+    - name: Take screenshot
+      if: failure()
+      continue-on-error: true
+      shell: >-
+        sudo --user=ci-user --login --set-home --non-interactive
+        bash --noprofile --norc -eo pipefail {0}
+      run: |
+        set -o xtrace -o errexit
+        PID=$(pidof smoke-test.sh || echo missing)
+        if [[ ! -r /proc/$PID/environ ]]; then
+          echo "Rancher Desktop is not running" >&2
+          exit 0
+        fi
+        export $(gawk 'BEGIN { RS="\0"; FS="=" } ($1 == "DISPLAY" || $1 == "XAUTHORITY") { print }' \
+          < /proc/$PID/environ)
+        env
+        export MAGICK_DEBUG=All # spellcheck-ignore-line
+        gm import -window root -verbose $LOGS_DIR/screenshot-${{ matrix.id }}.png
+
+    - name: Upload logs
+      uses: actions/upload-artifact@ea165f8d65b6e75b540449e92b4886f43607fa02 # v4.6.2
```

Comment:
Doesn't `actions/upload-artifact@v4` use the latest release, which would be `v4.6.2`?

Reply:
Yes, that's the [commit for v4.6.2](https://github.com/actions/upload-artifact/commit/ea165f8d65b6e75b540449e92b4886f43607fa02)?

At some point in the past, some other action got compromised and somebody pushed malicious code to existing tags (overwriting them). So now "best practice"† is to pin to the commit hash with a comment on the version, and dependabot knows how to update that.  Of course, that means it's pretty easy to sneak in a _different_ hash (can even be from a fork!) and compromise things, but one repo at a time…

You'll see that all of our other uses of `action/upload-artifact` does the same (I actually just copied it from an existing one).

† I kinda hate that phrase, or rather whenever it gets used since it's got a hint of cargo-culting…